File: api-server/app/schemas/attendance.py
from datetime import date

from pydantic import BaseModel, Field


class AttendanceCreate(BaseModel):
    game_id: int
    attend_type: str = "stadium"
    is_neutral: bool = False
    my_team_id: int | None = None
    seat_section: str | None = None
    seat_row: str | None = None
    seat_number: str | None = None
    memo: str | None = None
    rating: int | None = Field(default=None, ge=1, le=5)


class AttendanceUpdate(BaseModel):
    my_team_id: int | None = None
    seat_section: str | None = None
    seat_row: str | None = None
    seat_number: str | None = None
    memo: str | None = None
    rating: int | None = Field(default=None, ge=1, le=5)


class AttendanceRead(BaseModel):
    id: int
    user_id: int
    game_id: int
    attend_type: str
    my_team_id: int | None = None
    result_for_my_team: str | None = None
    seat_section: str | None = None
    seat_row: str | None = None
    seat_number: str | None = None
    memo: str | None = None
    rating: int | None = None
    game_date: date | None = None
    away_team_name: str | None = None
    home_team_name: str | None = None

    model_config = {"from_attributes": True}


class AttendanceBattingLeader(BaseModel):
    player_id: int
    player_name: str
    team_id: int
    team_name: str
    games: int
    pa: int
    h: int
    hr: int
    rbi: int
    bb: int
    obp: float
    slg: float
    ops: float


class AttendancePitchingLeader(BaseModel):
    player_id: int
    player_name: str
    team_id: int
    team_name: str
    games: int
    wins: int
    innings_pitched: float
    strikeouts: int
    era: float
    whip: float
    k_per_nine: float
    batting_average_against: float


class AttendanceDecisiveHitByGame(BaseModel):
    game_id: int
    game_date: date
    player_id: int | None = None
    player_name: str | None = None
    team_name: str | None = None


class AttendanceDecisiveHitLeader(BaseModel):
    player_id: int
    player_name: str
    team_name: str
    count: int


class AttendanceBreakdown(BaseModel):
    label: str
    games: int
    wins: int
    draws: int
    losses: int
    win_rate: float


class AttendanceSummaryRead(BaseModel):
    total_records: int
    qualified_games: int
    wins: int
    losses: int
    draws: int
    win_rate: float
    top_batting_players: list[AttendanceBattingLeader]
    top_pitchers: list[AttendancePitchingLeader]
    decisive_hit_leaders: list[AttendanceDecisiveHitLeader]
    decisive_hits: list[AttendanceDecisiveHitByGame]
    weekday_records: list[AttendanceBreakdown]
    stadium_records: list[AttendanceBreakdown]

File: mobile-app/lib/core/models/attendance.dart
class AttendanceModel {
  const AttendanceModel({
    required this.id,
    required this.gameId,
    required this.gameDate,
    required this.awayTeamName,
    required this.homeTeamName,
    this.awayTeamId,
    this.homeTeamId,
    this.myTeamId,
    this.resultForMyTeam,
    this.seatSection,
    this.memo,
  });

  factory AttendanceModel.fromJson(Map<String, dynamic> json) =>
      AttendanceModel(
        id: json['id'] as int,
        gameId: json['game_id'] as int,
        gameDate: DateTime.parse(json['game_date'] as String),
        awayTeamId: json['away_team_id'] as int?,
        awayTeamName: json['away_team_name'] as String? ?? '-',
        homeTeamId: json['home_team_id'] as int?,
        homeTeamName: json['home_team_name'] as String? ?? '-',
        myTeamId: json['my_team_id'] as int?,
        resultForMyTeam: json['result_for_my_team'] as String?,
        seatSection: json['seat_section'] as String?,
        memo: json['memo'] as String?,
      );

  final int id;
  final int gameId;
  final DateTime gameDate;
  final int? awayTeamId;
  final String awayTeamName;
  final int? homeTeamId;
  final String homeTeamName;
  final int? myTeamId;
  final String? resultForMyTeam;
  final String? seatSection;
  final String? memo;
}
